.github/workflows/ci-cd.yml

```
@@ -19,98 +19,106 @@ jobs:
-    needs: test
-    # Run only on pushing a tag
-    if: github.event_name == 'push' && contains(github.ref, 'refs/tags/')
-    steps:
-      - name: Dummy
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          submodules: true
+      - name: Setup Python ${{ matrix.pyver }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+          allow-prereleases: true
+      - name: Lock and sync dependencies
+        env:
+          AIOHTTP_VERSION: ${{ matrix.aiohttp-version }}
+        run: |
+          python -c "import sys; print(f'Python version: {sys.version}')"
+          python -m pip install -U pip codecov pip-tools
+          time pip-compile requirements-dev.in
+          time pip-sync requirements-dev.txt
+      - name: Run pre-commit hooks
+        if: matrix.python-version == '3.11'
+        run: |
+          make pre-commit
+      - name: Run unittests
+        env:
+          COLOR: 'yes'
         run: |
-            echo "Predeploy step"
+          make mototest
+      - name: Upload coverage to Codecov
+        if: matrix.python-version == '3.11'
+        uses: codecov/codecov-action@v4.5.0
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }} # not required for public repos
+          files: ./coverage.xml
+          flags: unittests # optional
+          name: codecov-umbrella # optional
+          fail_ci_if_error: true # optional (default = false)
+          verbose: true # optional (default = false)
 
-  build-tarball:
-    name: Tarball
+
+  # based on https://docs.github.com/en/actions/use-cases-and-examples/building-and-testing/building-and-testing-python#publishing-to-pypi
+  release-build:
     runs-on: ubuntu-latest
-    needs: pre-deploy
+
+    needs:
+      - test
+
     steps:
-    - name: Checkout
-      uses: actions/checkout@v4
-      with:
-        submodules: true
-    - name: Setup Python 3.8
-      uses: actions/setup-python@v5
-      with:
-        python-version: 3.8
-    - name: Make distributions (sdist and wheel)
-      run: |
-        python -c "import sys; print(f'Python version: {sys.version}')"
-        python -m pip install -U setuptools pip wheel
-        python setup.py sdist bdist_wheel
-    - name: Upload artifacts
-      uses: actions/upload-artifact@v4
-      with:
-        name: dist
-        path: dist
-
-  deploy:
-    name: Deploy
-    needs: [build-tarball]
+      - uses: actions/checkout@v4
+
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.8"
+
+      - name: Build release distributions
+        run: |
+          python -m pip install build
+          python -m build
+
+      - name: Check release distributions
+        run: |
+          python -m pip install twine
+          python -m twine check --strict dist/*
+
+      - name: Upload distributions
+        uses: actions/upload-artifact@v4
+        with:
+          name: release-dists
```

My guide https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/ uses
```suggestion
          name: python-package-distributions
```
because it represents what it is better. The possible use for these artifacts is wider. They are published on workflow runs and can be downloaded for inspection even when not releasing.
Additionally, if you get to testing them at some point, they'll be downloaded into the test job. So I'd not use the word “release” in the name.

P.S. Because they might end up being passed around into more places, I tend to use the name into a variable (like in the `${{ env }}` context) and just use that var everywhere instead of hardcoding the name in multiple places.